include/oneapi/dpl/internal/async_impl/async_utils.h

```diff
@@ -0,0 +1,187 @@
+/*
+ *  Copyright (c) 2020 Intel Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef _ONEDPL_ASYNC_UTILS_H
+#define _ONEDPL_ASYNC_UTILS_H
+
+#include <CL/sycl.hpp>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __internal
+{
+
+using sycl::event;
+
+template <typename _T, typename _Tp = sycl_iterator<sycl::access::mode::read_write, _T, sycl::buffer_allocator>>
```

Comment:
overall. I don't understand why we should mix generic programming and virtual functions. I will try to have closer look but at first glance virtual functions can be avoided.